File: maximum-product-of-splitted-binary-tree/maximum-product-of-splitted-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    static int MOD = (int)(1e9 + 7);
    
    int nodesSum;
    long maximumProduct;
    public int maxProduct(TreeNode root) {
        
        //find all the node sum
        nodesSum = getAllNodeSum(root);
        
        maximumProduct = 0l;
        
        //find the max product
        findMaxProduct(root);
        
        return (int)(maximumProduct % MOD);
    }
    
    private int getAllNodeSum(TreeNode node){
        if(node == null) return 0;
        
        return getAllNodeSum(node.left) + getAllNodeSum(node.right) + node.val;
    }
    
    
    private int findMaxProduct(TreeNode node){
        if(node == null) return 0;
        
        int substree1Sum = findMaxProduct(node.left) + findMaxProduct(node.right) + node.val;
        
        
        //remaining node, will be subtree 2 sum
        int substree2Sum = nodesSum - substree1Sum;
        
        maximumProduct = Math.max(maximumProduct, (long)substree1Sum * substree2Sum);
        
        return substree1Sum;
    }
}

File: 1465-maximum-product-of-splitted-binary-tree/maximum-product-of-splitted-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
long long total,ans;
const int mod=1e9+7;
    long long dfs(TreeNode* root){
        if(!root)return 0;
        return root->val+dfs(root->left)+dfs(root->right);
    }
    long long  dfs1(TreeNode* root){
        if(!root)return 0;
        int left_sum=dfs1(root->left);
        int right_sum=dfs1(root->right);
        int subarraysum=root->val+left_sum+right_sum;
        ans=max(ans,subarraysum*(total-subarraysum));
        return subarraysum;
    }
    int maxProduct(TreeNode* root) {
        total=dfs(root);
        ans=0;
       dfs1(root);
        ans=ans%mod;
       return (int)ans;
    }
};